software/SchemaTerms/sdoterm.py

```diff
@@ -17,65 +20,257 @@ class SdoTermType(str, enum.Enum):
     def __str__(self):
         return self.value
 
+class UnexpandedTermError(LookupError):
+    """Term is not expanded."""
+
+
+class SdoTermOrId(object):
+    """Wrapper that holds a term-id or a term."""
+
+    def __init__(self, term_id : str = None, term = None):
+        self._term = term
+        if term:
+            self._term_id = term.id
+            assert not term_id or term_id == term.id
+        else:
+            self._term_id = term_id
+
+    @property
+    def expanded(self) -> bool:
+        return not self._term_id or bool(self._term)
+
+    @property
+    def id(self) -> str:
+        return self._term_id
+
+    @property
+    def term(self):
+        if not self._term:
+             raise UnexpandedTermError()
+        return self._term
+
+    def setId(self, term_id):
+        self._term_id = term_id
+
+    def setTerm(self, term):
+        self._term = term
+        if self._term:
+            self._term_id = term.id
+
+    def __str__(self):
+        if not self._term:
+            return '<%s>' % self._term_id
+        return str(self._term)
+
+    def __bool__(self):
+        return bool(self._term_id)
+
+
+class SdoTermSequence(object):
+    """Sequence that holds either a sequence of term-ids, or a sequence of terms.
+
+    If it holds a sequence of terms, the sequence is said to expanded.
+    It can only be changed by replacing all the values, either with ids, or term instances.
+
+    """
+    def __init__(self):
+        self._term_dict = collections.OrderedDict()
+
+    @classmethod
+    def forElements(cls, elements):
+        """Convert an arbitrary sequence into a SdoTermSequence."""
+        if isinstance(elements, cls):
+            return elements
+
+        sequence = cls()
+        if all(map(lambda e : isinstance(e, SdoTerm), elements)):
+            sequence.setTerms(elements)
+            return sequence
+        term_ids = []
+        for element in elements:
+            try:
+                # This will work for both SdoTerm instances, and SdoTermOrId
+                term_id = element.id
+            except AttributeError:
+                term_id = element
+            term_ids.append(term_id)
+        sequence.setIds(term_ids)
+        return sequence
+
+    @property
+    def expanded(self):
+        return not self._term_dict or all(self._term_dict.values())
```

Comment:
`all` returns `True` when the iterable is empty, and the iterable is set in `__init__` so the first part of the test is not needed?

Reply:
Done.